Use hashes in Key instead of allocated strings

udoprog opened the pull request:
#### Description

Changes Key to wrap an `Rc<[u8]>` instead of an `Rc<str>`. Which allows `Key` to among other things implement `From<&[u8]>`. This in turn allows it to more efficiently wrap types which directly have a binary representation.

This does add some **unsafe** which we wouldn't have to if `From<Rc<str>>` existed in std, see rust-lang/rust#96078

This also introduces a handful of breaking changes which needs to be considered, especially since the binary representation of seemingly unrelated things might *overlap* or not:
* Integer types have lengths, so `0u32` would never match `0u64`. Previously this was slated over since they both compared to the string `"0"`.
* Different types can conflict if they have the same representation. Such as a `Vec<u8>` which has the same byte sequence as a `String`. Like `"foo"` and `b"foo"`.

In my mind the shortcomings are mitigated since *keys tend to be uniform* in the context which they are used. Consider when you're iterating over something and generating keys, they tend to be generated in the same manner:

```rust
html! {
    {for entries.iter().map(|e| {
        html! { <Entry key={e.id} entry={e} /> }
    }}
}
```

The upside of this is that using keys is made more efficient and sometimes avoids the extra allocation necessary to *stringify* things. My motivating use-case was that I wanted to use a [`Uuid`] as `key`. With this change I could directly use [`Uuid::as_bytes`].

[`Uuid`]: https://docs.rs/uuid/latest/uuid/struct.Uuid.html
[`Uuid::as_bytes`]: https://docs.rs/uuid/latest/uuid/struct.Uuid.html#method.as_bytes

#### Solving the shortcomings

If we want to, I do believe we could solve the above shortcomings with a bit of extra work.

* We can store the [TypeId](https://doc.rust-lang.org/std/any/struct.TypeId.html) in the `Key` or define our own. E.g. I personally think it would be reasonable to compare Rust strings with their binary equivalents since UTF-8 is intentionally ascii-compatible.
* We can use a [variable-length encoding](https://en.wikipedia.org/wiki/Variable-length_quantity) for integers, which would have the same binary representation regardless of size. If all numbers utilized [zig-zag encoding](https://en.wikipedia.org/wiki/Variable-length_quantity#Zigzag_encoding), signed and unsigned integers could be correctly compared.

#### Checklist

<!-- For further details, please read CONTRIBUTING.md -->

- [x] I have run `cargo make pr-flow`
- [x] I have reviewed my own code
- [x] I have added tests
  <!-- If this is a bug fix, these tests will fail if the bug is present (to stop it from cropping up again) -->
  <!-- If this is a feature, my tests prove that the feature works -->

futursolo commented:
Instead of converting internal representation of keys from `Rc<str>` to `Rc<[u8]>`, how about implementing it with `std::hash::Hash`?

hamza1311 commented:
If it's done with Hash, it's worth benchmarking the performance (not sure how good our CI is for that, @voidpumpkin can comment on that). While I doubt there will be performance impact, we may see a noticeable increase in binary size by adding another dependency

futursolo commented:
> we may see a noticeable increase in binary size by adding another dependency

We can use `std::collections::hash_map::DefaultHasher`?

udoprog commented:
@hamza1311 Tried a branch which uses hashing (but does not use zig-zag encoding): https://github.com/udoprog/yew/commit/key-hashing

Size differences when cherry picked on top of 0.19.3 and used in one of my applications (twox-hash added as a dependency):
```text
before:           10934052
after:            11155539
diff:               221487

before optimized:  1674539
after optimized:   1715264
diff:                40725
```

@futursolo It's an option. You'd have to be comfortable with 64-bit wide keys and it does do a bit of extra work by virtue of being cryptographic and randomly seeded. But I don't think that matters much.

hamza1311 commented:
I assume that is in bytes? An increase of ~40KB is a significant increase in binary size

udoprog commented:
I replaced the current propsal with one that does hashing, and `twox-hash` with 128-bit keys is hidden behind a feature flag w/ the same name. This is the result from the benches I added for anyone curious:

#### Current Key:
```
test bench_string_key ... bench:       3,385 ns/iter (+/- 91)
test bench_u128_key   ... bench:       7,589 ns/iter (+/- 218)
test bench_u64_key    ... bench:       7,589 ns/iter (+/- 469)
```
#### DefaultHasher (using 64-bit keys):
```
test bench_string_key ... bench:       3,020 ns/iter (+/- 99)
test bench_u128_key   ... bench:       1,788 ns/iter (+/- 38)
test bench_u64_key    ... bench:       1,781 ns/iter (+/- 20)
```
#### twox-hash (using 128-bit keys):
```
test bench_string_key ... bench:       6,581 ns/iter (+/- 3,075)
test bench_u128_key   ... bench:       6,500 ns/iter (+/- 86)
test bench_u64_key    ... bench:       6,494 ns/iter (+/- 326)
```

udoprog commented:
> I assume that is in bytes? An increase of ~40KB is a significant increase in binary size

Up to you to decide. I get something like 10kb variability between builds so it isn't that big of a deal for me. But I put twox-hash behind a non-default feature flag in this proposal as a result of your comment. If you don't want it at all, please tell me and I'll remove it.

WorldSEnder commented:
> I get something like 10kb variability between builds so it isn't that big of a deal for me.

It sounds we might use different measurement processes. There is a CI test for size comparison, and any change above ~30 bytes is out of the ordinary if functionality is unchanged. Be sure to run the comparison with build flags optimized for size (see the [size-cmp workflow](https://github.com/yewstack/yew/blob/c5b77909496333d421b5ee6b76024a8800e52bc7/.github/workflows/size-cmp.yml#L60-L71)) which should run wasm-opt and use nightly rust for tree-shaking and best size performance.

lukechu10 commented:
What about hash collisions? The std implementation of `HashMap` has special code for dealing with hash collisions which can be seen here: https://docs.rs/hashbrown/latest/src/hashbrown/raw/mod.rs.html#822

Although hash collisions should be relatively rare, especially with the default hasher, they still can occur. And the biggest problem that I see with this is that hashes are not deterministic which means that an error caused by a hash collision would not be reproducible.

udoprog commented:
> What about hash collisions? The std implementation of `HashMap` has special code for dealing with hash collisions which can be seen here: https://docs.rs/hashbrown/latest/src/hashbrown/raw/mod.rs.html#822

Hash collisions in a HashMap are common because the hash is used to pick between a much lower number of buckets, which are sized according to the capacity of the hash map.

The way we'd use keys and given that the a hash has a decent distribution (which I believe `DefaultHash` is) if we want to keep the probability of a single collision below 0.000001% we can have:
* 6.1M elements with a 64-bit hash.
* 6e16 elements with a 128-bit hash.

If I'm reading the table here correctly: <https://en.wikipedia.org/wiki/Birthday_attack>

I prefer the latter (which is why I at least want it as a feature to test it). But I don't see much of an issue with the former which is why I also think a 64-bit hash is fine. Random UUIDs for example are 128-bit which is considered sufficient to be globally unique.

> Although hash collisions should be relatively rare, especially with the default hasher, they still can occur. And the biggest problem that I see with this is that hashes are not deterministic which means that an error caused by a hash collision would not be reproducible.

The hashes proposed here are deterministic. Both `DefaultHasher` and `twox-hash` here are initialized with fixed seeds.

hamza1311 commented:
> I get something like 10kb variability between builds so it isn't that big of a deal for me

In release builds with `wasm-opt` and the [build size optimizations](https://yew.rs/docs/next/advanced-topics/optimizations)? If so, that seems a little off to me.

WorldSEnder commented:
> Furthermore, is that or anything else blocking this PR? Is this something you want to adopt?

I see two blockers, one you can't do anything about. I'd want to run the performance benchmarks locally, to see how exactly it affects them, since the github runner is a bit unstable and flactuates too much for accurate results.

The other things is clippy complaining about the use of `#![features(test)]` in the benchmarks. Can you change it, so that the benchmarks only get picked up on nightly?

udoprog commented:
Switched to criterion, since it's way harder to just selectively disable the default benches.

@WorldSEnder If I read you correct you personally wanted to run the webdriver benchmarks?

WorldSEnder commented:
> @WorldSEnder If I read you correct you personally wanted to run the webdriver benchmarks?

Yep. Running the `benchmark-struct` and `benchmark-hooks` tests locally for accurate results on multiple envs (chrome/firefox)*2 different CPUs.

WorldSEnder commented:
Update: looks like wins across the board. I think it's definitely something for the release profile and it has some nice performance wins on update. [Measurements](https://gist.github.com/WorldSEnder/c627f167d694261b81d1410beaaea7db). There seems to be a persistent small overhead when creating "trivial" keys. Which brings me to three more questions I want to ask:

- [ ] Can this benefit from niche optimization? Specifically, we use `Option<Key>` a few times in the code. Can we reserve *one* hash value as a niche such that `size_of::<Option<Key>> == size_of::<Key>`?
- [ ] Do we have to use such a *good* hashing algorithm, or could we stick in something simple like [`FxHasher`](https://docs.rs/fxhash/0.2.1/fxhash/struct.FxHasher.html), which should be even faster, especially when the input is a single `u64` getting hashed.
- [ ] Should we include the original key during debugging to provide a bit of a safety-net against hash collisions for good measure? In any case, we might want to make it clear *in the documentation* that the key should not be initialized with user/content data, to be a bit more resistant against crafted collisions.

udoprog commented:
> Can this benefit from niche optimization? Specifically, we use Option<Key> a few times in the code. Can we reserve one hash value as a niche such that size_of::<Option<Key>> == size_of::<Key>?

I suppose we can add a loop which tries to construct a NonZeroU64/128. Alternatively provide a default constructor which returns an empty all zero key that can be used instead. Does seem like a good follow-up PR to me.

> Do we have to use such a good hashing algorithm, or could we stick in something simple like FxHasher, which should be even faster, especially when the input is a single u64 getting hashed.

FxHash is in my mind is very similar to a single round of fnv, which does not provide very randomly distributed outputs making it more prone to conflict. How much? Not sure. Increasing the number of rounds helps, but then it's supposedly much slower than something like xxhash (or siphash).

> the key should not be initialized with user/content data, to be a bit more resistant against crafted collisions.

I'm not following. Could you elaborate? The proposed algorithms all have good avalanche effects, which should mean that crafting colissions shouldn't be easy.

WorldSEnder commented:
> I'm not following. Could you elaborate?

Sorry, I bunched a few things together. The crafted part was conditional on using a faster but worse hasher such as Fx. The overall recommendation still stands though, especially when using u64 hashes. A chance of 1e-9 for ~6.4k element *will* happen if you start testing with actually random data, and it will be very hard to debug. If you use controlled keys, such as a running index in a collection, the debuggability should be a lot higher, and tying in with testing against the full, unhashed, key in debug mode, should give at least reproducible issues if they do happen.

futursolo commented on packages/yew/src/virtual_dom/key.rs:
```diff
@@ -1,74 +1,189 @@
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        self.key.fmt(f)
+// Default implementation for byte sequences.
+impl From<&[u8]> for Key {
```
Why not `impl<T> From<T> for Key where T: Hash` and only specialise on types that needs special handling?

udoprog replied:
Providing a blanket `impl<T> From<T> for Key where T: Hash` would prevent any specialization, since `Hash` might be implemented upstream for any `T` in the future (trait coherence rules) and we don't have access to stable `specialization.

futursolo replied:
I guess instead of using `From<T>`, having `pub fn new<H: Hash>(h: H) -> Key` as an associated function would work.

Usage of `From`(Into) for conversion into Keys is designated by the `html!` macro and can be changed to something else.

WorldSEnder replied:
I don't see how it semantically makes sense to construct a `Key` from an abstract hashable type. I see the hashing as a performance improvement that acts as-if it would compare the actual values the `Key` was constructed from. I wouldn't expand this. Users can just hash those types themselves before constructing the keys and wrestle their own collision problems.

futursolo commented on packages/yew/src/virtual_dom/key.rs:
```diff
@@ -1,74 +1,189 @@
-key_impl_from_to_string!(i64);
-key_impl_from_to_string!(i128);
-key_impl_from_to_string!(isize);
+macro_rules! signed {
```
Does this mean that `Key::from(0_u8) == Key::from(0_u16)` but `Key::from(0_i8) != Key::from(0_u8)`?

udoprog replied:
Yeah, this can be remedied with zig zag encoding [as mentioned under **Solving the shortcomings** here](https://github.com/yewstack/yew/pull/2616#issue-1205585356).

futursolo replied:
Personally, I feel that passing a key of `0_u16` into an element / component where its last key is `0_u8` to be treated as the same element / component may not be an intentional behaviour for most cases.

I would prefer to treat them as different values if a correct implementation could potentially negatively impact code size.

We can emit a warning if the type of key changes between renders in debug mode.

futursolo reviewed (commented):
Looks good for the most part, just some minor issues.

In addition, we should also look into why the code size of `function_router` has increased for 8KB.

udoprog commented:
Uh, as a heads up I didn't intend to close it. I did remove my fork while not remembering this was still open. But it didn't seem like anyone moved on this anyway.
